File: mathLog/TaskB/src/parser/tokens/Tokenizer.java
package parser.tokens;

public class Tokenizer {
    private final String expression;
    private Tokens token = Tokens.START;
    private int ind = -1;
    private String lastVar;

    public Tokenizer(String expression) {
        this.expression = expression;
    }

    public void skipWhiteSpace() {
        while (Character.isWhitespace(expression.charAt(ind))) {
            ind++;
            if (ind >= expression.length()) {
                token = Tokens.END;
                break;
            }
        }

    }

    public void nextToken() {
        ind++;
        if (ind < expression.length())
            skipWhiteSpace();
        else
            token = Tokens.END;

        if (token == Tokens.END)
            return;

        switch (expression.charAt(ind)) {
            case '(':
                token = Tokens.OPEN_BRACKET;
                break;
            case ')':
                token = Tokens.CLOSE_BRACKET;
                break;
            case '|':
                token = Tokens.OP_OR;
                break;
            case '&':
                token = Tokens.OP_AND;
                break;
            case '!':
                token = Tokens.OP_NEG;
                break;
            case '→':
                token = Tokens.OP_FOLLOW;
                break;
            case '⊢':
                token = Tokens.TURNSTILE;
                break;
            default:
                token = Tokens.VAR;
                parseVar();
                break;
        }
    }

    public void parseVar() {
        var res = new StringBuilder();
        char symb = expression.charAt(ind);

        while ((symb >= 'A' && symb <= 'Z') || (symb >= '0' && symb <= '9') || (symb == '\'')) {
            res.append(symb);
            ind++;
            if (ind >= expression.length()) {
                break;
            }
            symb = expression.charAt(ind);
        }
        ind--;
        lastVar = res.toString();
    }

    public Tokens getToken() {
        return token;
    }

    public String getLastVar() {
        return lastVar;
    }
}

File: android/Hw1_calculator/app/src/main/java/com/lab/hw1/parser/tokenizer/Tokenizer.kt
package com.lab.hw1.parser.tokenizer

import com.lab.hw1.parser.exceptions.ParsingException


class Tokenizer(private val expression: String) {
    private var ind: Int = -1
    var lastConst: Double = 0.0
        private set
    var token: Token = Token.START
        private set

    private fun skipWhiteSpaces() {
        while (ind < expression.length && Character.isWhitespace(expression[ind])) {
            ind++
        }
    }

    private fun parseConst(): Double {
        val addDigits = { argThis: StringBuilder ->
            while (ind < expression.length && expression[ind].isDigit()) {
                argThis.append(expression[ind])
                ind++
            }
        }
        val res = StringBuilder()
        addDigits.invoke(res)
        if (ind < expression.length && expression[ind] == '.') {
            res.append(expression[ind])
            ind++
            addDigits.invoke(res)
        }
        ind--
        return res.toString().toDouble()
    }

    fun nextToken() {
        ind++
        skipWhiteSpaces()

        if (ind >= expression.length) {
            token = Token.END
            return
        }

        when (expression[ind]) {
            '(' -> token = Token.OPEN_BR
            ')' -> token = Token.CLOSE_BR
            '+' -> token = Token.PLUS
            '-' -> token = Token.MINUS
            '*' -> token = Token.MUL
            '/' -> token = Token.DIV

            else -> {
                if (Character.isDigit(expression[ind])) {
                    lastConst = parseConst()
                    token = Token.CONST
                } else
                    throwException("unexpected symbol")
            }
        }
    }

    fun throwException(message: String) {
        val t =
            if (ind + 1 < expression.length)
                ind + 1
            else
                expression.length

        throw ParsingException(
            expression.substring(0, t) + "...| error: $message",
            ind
        )
    }
}